File: benchmarks/nyc_taxi/r/q12_multi_stage_rollup.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 1) {
  stop("usage: Rscript q12_multi_stage_rollup.R <parquet-dataset-path>", call. = FALSE)
}

suppressPackageStartupMessages({
  library(arrow)
  library(dplyr)
})

started <- Sys.time()
dataset_path <- args[[1]]
ds_tbl <- open_dataset(dataset_path, format = "parquet")

result <- ds_tbl |>
  filter(!is.na(trip_distance), !is.na(fare_amount), trip_distance > 0, total_amount > 0) |>
  mutate(
    fare_per_mile = fare_amount / trip_distance,
    tip_percent = (tip_amount / total_amount) * 100
  ) |>
  group_by(year, month, VendorID, payment_type) |>
  summarise(
    avg_fare_per_mile = mean(fare_per_mile, na.rm = TRUE),
    avg_tip_percent = mean(tip_percent, na.rm = TRUE),
    total_revenue = sum(total_amount, na.rm = TRUE),
    max_trip_distance = max(trip_distance, na.rm = TRUE),
    rides = n(),
    unique_pickups = n_distinct(PULocationID)
  ) |>
  arrange(desc(total_revenue)) |>
  collect()

print(head(result, 10))
cat("ROWS_SCANNED=NA\n")
cat(sprintf("ROWS_RETURNED=%s\n", nrow(result)))
cat(sprintf("ELAPSED_SEC=%.6f\n", as.numeric(difftime(Sys.time(), started, units = "secs"))))

File: benchmarks/nyc_taxi/python/q12_multi_stage_rollup.py
import argparse
import time

import pyarrow.dataset as ds


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--dataset", required=True, help="Path to the partitioned parquet dataset")
    args = parser.parse_args()

    started = time.perf_counter()
    dataset = ds.dataset(args.dataset, format="parquet", partitioning="hive")

    table = dataset.to_table(
        columns=[
            "year",
            "month",
            "VendorID",
            "payment_type",
            "PULocationID",
            "trip_distance",
            "fare_amount",
            "tip_amount",
            "total_amount",
        ]
    )
    frame = table.to_pandas()
    frame = frame[
        frame["trip_distance"].notna()
        & frame["fare_amount"].notna()
        & (frame["trip_distance"] > 0)
        & (frame["total_amount"] > 0)
    ].copy()
    frame["fare_per_mile"] = frame["fare_amount"] / frame["trip_distance"]
    frame["tip_percent"] = (frame["tip_amount"] / frame["total_amount"]) * 100.0
    result = (
        frame.groupby(["year", "month", "VendorID", "payment_type"], dropna=False)
        .agg(
            avg_fare_per_mile=("fare_per_mile", "mean"),
            avg_tip_percent=("tip_percent", "mean"),
            total_revenue=("total_amount", "sum"),
            max_trip_distance=("trip_distance", "max"),
            rides=("VendorID", "size"),
            unique_pickups=("PULocationID", "nunique"),
        )
        .reset_index()
        .sort_values("total_revenue", ascending=False)
    )

    print(result.head(10))
    print("ROWS_SCANNED=NA")
    print(f"ROWS_RETURNED={len(result.index)}")
    print(f"ELAPSED_SEC={time.perf_counter() - started:.6f}")


if __name__ == "__main__":
    main()
